File: ProblemSolve/538.cs
/*******************************
 * Convert BST to Greater Tree *
 *******************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private int greaterSum = 0;
    public TreeNode ConvertBST(TreeNode root) {
        if(root == null){
            return null;
        }
        
        ConvertBST(root.right);
        greaterSum += root.val;
        root.val = greaterSum;
        ConvertBST(root.left);
        
        return root;
    }
}

File: ProblemSolve/538.cpp
/*******************************
 * Convert BST to Greater Tree *
 *******************************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int greaterSum = 0;
    
    TreeNode* convertBST(TreeNode* root) {
        if(root == NULL)
            return NULL;
        
        convertBST(root -> right);
        greaterSum += root -> val;
        root -> val = greaterSum;
        convertBST(root -> left);
        
        return root;
    }
};